chainer/links/connection/gru.py

```diff
@@ -56,3 +68,74 @@ def __call__(self, h, x):
         h_bar = tanh.tanh(self.W(x) + self.U(r * h))
         h_new = (1 - z) * h + z * h_bar
         return h_new
+
+
+class StatefulGRU(GRUBase):
+    """Stateful Gated Recurrent Unit function (GRU).
+
+    Stateful GRU function has six parameters :math:`W_r`, :math:`W_z`,
+    :math:`W`, :math:`U_r`, :math:`U_z`, and :math:`U`.
+    All these parameters are :math:`n \\times n` matricies,
+    where :math:`n` is the dimension of hidden vectors.
+
+    Given input vector :math:`x`, Stateful GRU returns the next
+    hidden vector :math:`h'` defined as
+
+    .. math::
+
+       r &=& \\sigma(W_r x + U_r h), \\\\
+       z &=& \\sigma(W_z x + U_z h), \\\\
+       \\bar{h} &=& \\tanh(W x + U (r \\odot h)), \\\\
+       h' &=& (1 - z) \\odot h + z \\odot \\bar{h},
+
+    where :math:`h` is current hidden vector.
+
+    As the name indicates, :class:`~chainer.links.StatefulGRU` is *stateful*,
+    meaning that it also holds the next hidden vector `h'` as a state.
+    Use :class:`~chainer.links.GRU` as a stateless version of GRU.
+
+    Args:
+        n_units(int): Dimension of input vector :math:`x`, and hidden vector.
+
+    Attributes:
+        h(~chainer.Variable): Hidden vector that indicates the state of
+        :class:`~chainer.links.StatefulGRU`.
+
+    .. seealso:: :class:`~chainer.functions.GRU`
+    """
+
+    def __init__(self, n_units):
+        super(StatefulGRU, self).__init__(n_units)
+        self.state_size = n_units
+        self.reset_state()
+
+    def to_cpu(self):
+        super(StatefulGRU, self).to_cpu()
+        if self.h is not None:
+            self.h.to_cpu()
+
+    def to_gpu(self, device=None):
+        super(StatefulGRU, self).to_gpu(device)
+        if self.h is not None:
+            self.h.to_gpu(device)
+
+    def set_state(self, h):
+        if self.xp == numpy:
+            h = chainer.cuda.to_cpu(h)
+        else:
+            h = chainer.cuda.to_gpu(h)
+        self.h = chainer.Variable(h)
+
+    def reset_state(self):
+        self.h = None
+
+    def __call__(self, x):
+        if self.h is None:
```

Comment:
If self.h is None, you can omit the lateral computations (`U_r(h)`, `U_z(h)`, `U(r*h)`, and `(1-z)*h`) instead of initializing it by zeros. It would slightly improve the first iteration on a sequence.